File: ios/HealthTracker/Utils/DemoDataGenerator.swift
import Foundation
import CoreData

class DemoDataGenerator {
    static func generateDemoData(context: NSManagedObjectContext) {
        // Create demo user profile if needed
        createDemoUserProfile()
        
        // Clear existing demo data first
        clearExistingDemoData(context: context)
        
        // Generate data for the past 30 days
        let today = Date()
        
        // Generate food entries
        generateFoodEntries(context: context, startDate: today)
        
        // Generate exercise entries
        generateExerciseEntries(context: context, startDate: today)
        
        // Generate weight entries
        generateWeightEntries(context: context, startDate: today)

        // Generate water entries
        generateWaterEntries(context: context, startDate: today)

        // Generate supplement entries
        generateSupplementEntries(context: context, startDate: today)

        // Set demo step count in UserDefaults for today
        setDemoStepCount()

        // Generate achievements
        generateAchievements()
        
        // Generate meal plans
        generateMealPlans()
        
        // Save all changes
        do {
            try context.save()
            print("Demo data generated successfully!")
        } catch {
            print("Error saving demo data: \(error)")
        }
    }
    
    private static func clearExistingDemoData(context: NSManagedObjectContext) {
        // Clear food entries
        let foodRequest: NSFetchRequest<NSFetchRequestResult> = FoodEntry.fetchRequest()
        let foodDelete = NSBatchDeleteRequest(fetchRequest: foodRequest)
        
        // Clear exercise entries
        let exerciseRequest: NSFetchRequest<NSFetchRequestResult> = ExerciseEntry.fetchRequest()
        let exerciseDelete = NSBatchDeleteRequest(fetchRequest: exerciseRequest)
        
        // Clear weight entries
        let weightRequest: NSFetchRequest<NSFetchRequestResult> = WeightEntry.fetchRequest()
        let weightDelete = NSBatchDeleteRequest(fetchRequest: weightRequest)
        
        // Clear supplement entries
        let supplementRequest: NSFetchRequest<NSFetchRequestResult> = SupplementEntry.fetchRequest()
        let supplementDelete = NSBatchDeleteRequest(fetchRequest: supplementRequest)
        
        do {
            try context.execute(foodDelete)
            try context.execute(exerciseDelete)
            try context.execute(weightDelete)
            try context.execute(supplementDelete)
        } catch {
            print("Error clearing existing data: \(error)")
        }
    }
    
    private static func generateFoodEntries(context: NSManagedObjectContext, startDate: Date) {
        let calendar = Calendar.current
        
        // Sample meals for variety
        let breakfastOptions = [
            ("Oatmeal with Berries", "Quaker", 250.0, 8.0, 45.0, 5.0, 8.0),
            ("Greek Yogurt Parfait", "Chobani", 320.0, 20.0, 35.0, 12.0, 4.0),
            ("Scrambled Eggs with Toast", nil, 380.0, 22.0, 28.0, 18.0, 2.0),
            ("Protein Smoothie", "Homemade", 295.0, 25.0, 40.0, 6.0, 5.0),
            ("Avocado Toast", nil, 340.0, 10.0, 36.0, 18.0, 8.0)
        ]
        
        let lunchOptions = [
            ("Grilled Chicken Salad", "Sweetgreen", 420.0, 35.0, 25.0, 22.0, 6.0),
            ("Turkey Sandwich", "Subway", 380.0, 24.0, 46.0, 12.0, 3.0),
            ("Quinoa Buddha Bowl", nil, 485.0, 18.0, 65.0, 18.0, 12.0),
            ("Salmon with Vegetables", nil, 450.0, 40.0, 20.0, 25.0, 5.0),
            ("Chicken Burrito Bowl", "Chipotle", 680.0, 32.0, 65.0, 32.0, 8.0)
        ]
        
        let dinnerOptions = [
            ("Grilled Salmon with Rice", nil, 520.0, 42.0, 48.0, 18.0, 2.0),
            ("Pasta with Chicken", "Homemade", 580.0, 38.0, 62.0, 20.0, 4.0),
            ("Steak with Sweet Potato", nil, 650.0, 45.0, 42.0, 32.0, 6.0),
            ("Vegetarian Curry", "Thai Kitchen", 420.0, 12.0, 58.0, 16.0, 8.0),
            ("Shrimp Stir-fry", nil, 380.0, 28.0, 45.0, 10.0, 6.0)
        ]
        
        let snackOptions = [
            ("Apple with Almond Butter", nil, 200.0, 4.0, 25.0, 10.0, 4.0),
            ("Protein Bar", "RXBAR", 210.0, 12.0, 23.0, 9.0, 3.0),
            ("Mixed Nuts", "Planters", 180.0, 6.0, 8.0, 16.0, 2.0),
            ("Banana", nil, 105.0, 1.3, 27.0, 0.4, 3.1),
            ("Greek Yogurt", "Fage", 130.0, 16.0, 9.0, 4.0, 0.0)
        ]
        
        // Generate entries for past 30 days
        for dayOffset in 0..<30 {
            guard let date = calendar.date(byAdding: .day, value: -dayOffset, to: startDate) else { continue }
            
            // Breakfast
            let breakfast = breakfastOptions.randomElement()!
            createFoodEntry(context: context, date: date, mealType: .breakfast,
                          name: breakfast.0, brand: breakfast.1, calories: breakfast.2,
                          protein: breakfast.3, carbs: breakfast.4, fat: breakfast.5, fiber: breakfast.6)
            
            // Lunch
            let lunch = lunchOptions.randomElement()!
            createFoodEntry(context: context, date: date, mealType: .lunch,
                          name: lunch.0, brand: lunch.1, calories: lunch.2,
                          protein: lunch.3, carbs: lunch.4, fat: lunch.5, fiber: lunch.6)
            
            // Dinner
            let dinner = dinnerOptions.randomElement()!
            createFoodEntry(context: context, date: date, mealType: .dinner,
                          name: dinner.0, brand: dinner.1, calories: dinner.2,
                          protein: dinner.3, carbs: dinner.4, fat: dinner.5, fiber: dinner.6)
            
            // Snacks (1-2 per day)
            let snackCount = Int.random(in: 1...2)
            for _ in 0..<snackCount {
                let snack = snackOptions.randomElement()!
                createFoodEntry(context: context, date: date, mealType: .snack,
                              name: snack.0, brand: snack.1, calories: snack.2,
                              protein: snack.3, carbs: snack.4, fat: snack.5, fiber: snack.6)
            }
        }
    }
    
    private static func createFoodEntry(context: NSManagedObjectContext, date: Date, mealType: MealType,
                                      name: String, brand: String?, calories: Double,
                                      protein: Double, carbs: Double, fat: Double, fiber: Double) {
        let entry = FoodEntry(context: context)
        entry.id = UUID()
        entry.name = name
        entry.brand = brand
        entry.calories = calories
        entry.protein = protein
        entry.carbs = carbs
        entry.fat = fat
        entry.fiber = fiber
        entry.mealType = mealType.rawValue
        entry.timestamp = date
        entry.date = date
        entry.servingSize = "1"
        entry.servingUnit = "serving"
    }
    
    private static func generateExerciseEntries(context: NSManagedObjectContext, startDate: Date) {
        let calendar = Calendar.current
        
        let exercises = [
            ("Morning Run", "Cardio", 30, 280.0),
            ("Weight Training", "Strength", 45, 220.0),
            ("Yoga", "Flexibility", 60, 180.0),
            ("Cycling", "Cardio", 45, 400.0),
            ("Swimming", "Cardio", 30, 350.0),
            ("HIIT Workout", "Cardio", 20, 250.0),
            ("Pilates", "Strength", 40, 200.0),
            ("Walking", "Cardio", 60, 240.0)
        ]
        
        // Generate exercises for past 30 days (4-5 times per week)
        for dayOffset in 0..<30 {
            guard let date = calendar.date(byAdding: .day, value: -dayOffset, to: startDate) else { continue }
            
            // 70% chance of exercise on any given day
            if Double.random(in: 0...1) < 0.7 {
                let exercise = exercises.randomElement()!
                
                let entry = ExerciseEntry(context: context)
                entry.id = UUID()
                entry.name = exercise.0
                entry.category = exercise.1
                entry.type = exercise.1
                entry.duration = Int32(exercise.2)
                entry.caloriesBurned = exercise.3
                entry.timestamp = date
                entry.date = date
                entry.notes = "Felt great!"
            }
        }
    }
    
    private static func generateWeightEntries(context: NSManagedObjectContext, startDate: Date) {
        let calendar = Calendar.current
        let currentWeight = 170.0 // Current weight showing on dashboard
        let startingWeight = 180.0 // Starting weight 30 days ago
        let goalWeight = 160.0 // Goal weight

        // Generate weight entries every 2-3 days showing gradual loss
        var previousWeight = startingWeight
        for dayOffset in [28, 25, 22, 19, 16, 14, 11, 8, 5, 3, 0] {
            guard let date = calendar.date(byAdding: .day, value: -dayOffset, to: startDate) else { continue }

            // Calculate weight with gradual loss trend
            let progress = Double(28 - dayOffset) / 28.0
            let targetLoss = startingWeight - currentWeight
            let baseWeight = startingWeight - (targetLoss * progress)

            // Add small realistic fluctuation
            let fluctuation = dayOffset == 0 ? 0 : Double.random(in: -0.3...0.3)
            var weight = baseWeight + fluctuation

            // Make sure today's weight is exactly 170
            if dayOffset == 0 {
                weight = currentWeight
            }

            let entry = WeightEntry(context: context)
            entry.id = UUID()
            entry.weight = weight
            entry.timestamp = date
            entry.date = date

            // Add motivating notes
            if dayOffset == 0 {
                entry.notes = "New low! 10 lbs down 🎉"
            } else if dayOffset == 3 {
                entry.notes = "Almost at 170!"
            } else if dayOffset == 14 {
                entry.notes = "Halfway to goal weight"
            }

            previousWeight = weight
        }
    }
    
    private static func generateWaterEntries(context: NSManagedObjectContext, startDate: Date) {
        let calendar = Calendar.current

        // Generate water intake for today and past week
        for dayOffset in 0..<7 {
            guard let date = calendar.date(byAdding: .day, value: -dayOffset, to: startDate) else { continue }

            // Generate 6-8 water entries throughout the day
            let numEntries = Int.random(in: 6...8)
            let targetOunces = 64 // Daily goal

            for entryNum in 0..<numEntries {
                // Spread entries throughout the day
                let hour = 7 + (entryNum * 2) // Starting at 7am, every 2 hours
                guard let entryDate = calendar.date(bySettingHour: hour, minute: 0, second: 0, of: date) else { continue }

                let entry = WaterEntry(context: context)
                entry.id = UUID()
                entry.amount = Double.random(in: 8...12) // 8-12 oz per entry
                entry.unit = "oz"
                entry.timestamp = entryDate

                // For today, make sure we're at good hydration
                if dayOffset == 0 && entryNum < 5 {
                    entry.amount = 10 // Consistent 10 oz entries
                }
            }
        }
    }

    private static func generateSupplementEntries(context: NSManagedObjectContext, startDate: Date) {
        let calendar = Calendar.current
        
        let supplements = [
            ("Multivitamin", "Centrum", ["Vitamin A": 900.0, "Vitamin C": 90.0, "Vitamin D": 20.0, "Iron": 18.0]),
            ("Omega-3", "Nordic Naturals", ["EPA": 180.0, "DHA": 120.0]),
            ("Vitamin D3", "Nature Made", ["Vitamin D": 50.0]),
            ("Probiotics", "Garden of Life", ["CFU": 50.0]),
            ("Magnesium", "NOW Foods", ["Magnesium": 400.0])
        ]
        
        // Generate daily supplement entries
        for dayOffset in 0..<30 {
            guard let date = calendar.date(byAdding: .day, value: -dayOffset, to: startDate) else { continue }
            
            // Take 2-3 supplements per day
            let supplementCount = Int.random(in: 2...3)
            let selectedSupplements = supplements.shuffled().prefix(supplementCount)
            
            for supplement in selectedSupplements {
                let entry = SupplementEntry(context: context)
                entry.id = UUID()
                entry.name = supplement.0
                entry.brand = supplement.1
                entry.nutrients = supplement.2
                entry.timestamp = date
                entry.date = date
                entry.servingSize = "1"
                entry.servingUnit = "tablet"
            }
        }
    }
    
    private static func setDemoStepCount() {
        // Set a good step count for today (shown in dashboard)
        UserDefaults.standard.set(6915, forKey: "demoStepCount")
        UserDefaults.standard.set(Date(), forKey: "demoStepCountDate")
    }

    private static func generateAchievements() {
        let achievementManager = AchievementManager.shared
        
        // Add some demo achievements
        let achievements = [
            Achievement(
                type: .weightLoss,
                title: "5 lbs Down!",
                description: "You've lost 5 pounds!",
                dateEarned: Date().addingTimeInterval(-7 * 24 * 60 * 60), // 7 days ago
                value: 5.0,
                target: nil
            ),
            Achievement(
                type: .exerciseStreak,
                title: "7 Day Streak",
                description: "Exercised for 7 days in a row!",
                dateEarned: Date().addingTimeInterval(-2 * 24 * 60 * 60), // 2 days ago
                value: 7.0,
                target: nil
            ),
            Achievement(
                type: .calorieGoal,
                title: "Calorie Champion",
                description: "Met your calorie goal for 14 days!",
                dateEarned: Date().addingTimeInterval(-1 * 24 * 60 * 60), // Yesterday
                value: 14.0,
                target: nil
            ),
            Achievement(
                type: .waterIntake,
                title: "Hydration Hero",
                description: "Drank 8 glasses of water daily for a week!",
                dateEarned: Date().addingTimeInterval(-3 * 24 * 60 * 60), // 3 days ago
                value: 8.0,
                target: nil
            )
        ]
        
        // Add achievements to the manager
        for achievement in achievements {
            if !achievementManager.recentAchievements.contains(where: { $0.id == achievement.id }) {
                achievementManager.recentAchievements.append(achievement)
            }
        }
        
        // AchievementManager automatically saves when achievements are added
    }
    
    private static func createDemoUserProfile() {
        let userProfileManager = UserProfileManager()
        
        // Only create if no profile exists
        if userProfileManager.currentProfile == nil {
            var demoProfile = UserProfile(
                name: "Alex Johnson",
                gender: .male,
                birthDate: Calendar.current.date(byAdding: .year, value: -30, to: Date()) ?? Date()
            )
            
            // Set additional properties
            demoProfile.activityLevel = .moderate
            demoProfile.dietaryRestrictions = [DietaryRestriction.vegetarian]
            
            userProfileManager.currentProfile = demoProfile
            userProfileManager.hasCompletedOnboarding = true
        }
    }
    
    private static func generateMealPlans() {
        // Meal plans are now stored in Core Data, not UserDefaults
        // This function is no longer needed for demo data
    }
}

File: android/app/src/main/java/com/mochasmindlab/mlhealth/utils/DemoDataGenerator.kt
package com.mochasmindlab.mlhealth.utils

import com.mochasmindlab.mlhealth.data.entities.*
import com.mochasmindlab.mlhealth.data.models.FoodItem
import com.mochasmindlab.mlhealth.data.repository.*
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.random.Random

@Singleton
class DemoDataGenerator @Inject constructor(
    private val foodRepository: FoodRepository,
    private val exerciseRepository: ExerciseRepository,
    private val weightRepository: WeightRepository,
    private val waterRepository: WaterRepository,
    private val supplementRepository: SupplementRepository
) {

    suspend fun generateDemoData(daysBack: Int = 30) {
        generateFoodData(daysBack)
        generateExerciseData(daysBack)
        generateWeightData(daysBack)
        generateWaterData(daysBack)
        generateSupplementData(daysBack)
    }

    private suspend fun generateFoodData(daysBack: Int) {
        val calendar = Calendar.getInstance()
        val foods = getDemoFoods()

        for (day in 0 until daysBack) {
            calendar.time = Date()
            calendar.add(Calendar.DAY_OF_MONTH, -day)
            val date = calendar.time

            // Breakfast
            generateMealEntries(date, "BREAKFAST", foods.filter { it.category == "Breakfast" }, 1..2)

            // Lunch
            generateMealEntries(date, "LUNCH", foods.filter { it.category == "Lunch" }, 2..3)

            // Dinner
            generateMealEntries(date, "DINNER", foods.filter { it.category == "Dinner" }, 2..3)

            // Snacks
            generateMealEntries(date, "SNACK", foods.filter { it.category == "Snack" }, 0..2)
        }
    }

    private suspend fun generateMealEntries(
        date: Date,
        mealType: String,
        foods: List<DemoFood>,
        countRange: IntRange
    ) {
        val count = Random.nextInt(countRange.first, countRange.last + 1)
        val selectedFoods = foods.shuffled().take(count)

        selectedFoods.forEach { food ->
            val entry = FoodEntry(
                id = UUID.randomUUID().toString(),
                name = food.name,
                brand = food.brand,
                calories = food.calories,
                protein = food.protein,
                carbs = food.carbs,
                fat = food.fat,
                fiber = food.fiber ?: 0.0,
                sugar = food.sugar ?: 0.0,
                sodium = food.sodium ?: 0.0,
                servingSize = food.servingSize,
                servingUnit = food.servingUnit,
                servingCount = Random.nextDouble(0.5, 2.0).round(1),
                mealType = mealType,
                date = date,
                timestamp = date,
                barcode = null,
                imageUrl = null
            )
            foodRepository.addFoodEntry(entry)
        }
    }

    private suspend fun generateExerciseData(daysBack: Int) {
        val calendar = Calendar.getInstance()
        val exercises = getDemoExercises()

        for (day in 0 until daysBack) {
            calendar.time = Date()
            calendar.add(Calendar.DAY_OF_MONTH, -day)
            val date = calendar.time

            // 70% chance of exercise on any given day
            if (Random.nextDouble() < 0.7) {
                val exerciseCount = Random.nextInt(1, 4)
                val selectedExercises = exercises.shuffled().take(exerciseCount)

                selectedExercises.forEach { exercise ->
                    val duration = Random.nextInt(exercise.minDuration, exercise.maxDuration + 1)
                    val caloriesBurned = (exercise.caloriesPerMinute * duration).toInt()

                    val entry = ExerciseEntry(
                        id = UUID.randomUUID().toString(),
                        name = exercise.name,
                        category = exercise.category,
                        duration = duration,
                        caloriesBurned = caloriesBurned,
                        distance = if (exercise.hasDistance) Random.nextDouble(1.0, 10.0).round(1) else null,
                        distanceUnit = if (exercise.hasDistance) "km" else null,
                        notes = null,
                        date = date,
                        timestamp = date
                    )
                    exerciseRepository.addExerciseEntry(entry)
                }
            }
        }
    }

    private suspend fun generateWeightData(daysBack: Int) {
        val calendar = Calendar.getInstance()
        var currentWeight = 70.0 + Random.nextDouble(-5.0, 5.0) // Start around 70kg

        for (day in 0 until daysBack) {
            calendar.time = Date()
            calendar.add(Calendar.DAY_OF_MONTH, -day)
            val date = calendar.time

            // Weight entry every 2-3 days
            if (day % Random.nextInt(2, 4) == 0) {
                // Gradual weight change
                currentWeight += Random.nextDouble(-0.3, 0.2)

                val entry = WeightEntry(
                    id = UUID.randomUUID().toString(),
                    weight = currentWeight.round(1),
                    unit = "kg",
                    date = date,
                    timestamp = date,
                    bodyFatPercentage = if (Random.nextBoolean()) Random.nextDouble(15.0, 30.0).round(1) else null,
                    muscleMass = if (Random.nextBoolean()) Random.nextDouble(25.0, 40.0).round(1) else null,
                    notes = null
                )
                weightRepository.addWeightEntry(entry)
            }
        }
    }

    private suspend fun generateWaterData(daysBack: Int) {
        val calendar = Calendar.getInstance()

        for (day in 0 until daysBack) {
            calendar.time = Date()
            calendar.add(Calendar.DAY_OF_MONTH, -day)

            // Generate 4-10 water entries per day
            val entriesCount = Random.nextInt(4, 11)

            for (i in 0 until entriesCount) {
                calendar.set(Calendar.HOUR_OF_DAY, Random.nextInt(6, 22))
                calendar.set(Calendar.MINUTE, Random.nextInt(0, 60))

                val amounts = listOf(8f, 12f, 16f, 16.9f, 20f, 24f) // Common water amounts in oz

                val entry = WaterEntry(
                    id = UUID.randomUUID().toString(),
                    amount = amounts.random(),
                    unit = WaterUnit.OZ,
                    timestamp = calendar.time
                )
                waterRepository.addWaterEntry(entry)
            }
        }
    }

    private suspend fun generateSupplementData(daysBack: Int) {
        val calendar = Calendar.getInstance()
        val supplements = getDemoSupplements()

        for (day in 0 until daysBack) {
            calendar.time = Date()
            calendar.add(Calendar.DAY_OF_MONTH, -day)
            val date = calendar.time

            // 80% chance of taking supplements
            if (Random.nextDouble() < 0.8) {
                val selectedSupplements = supplements.shuffled().take(Random.nextInt(1, 4))

                selectedSupplements.forEach { supplement ->
                    val entry = SupplementEntry(
                        id = UUID.randomUUID().toString(),
                        name = supplement.name,
                        brand = supplement.brand,
                        dosage = supplement.dosage,
                        unit = supplement.unit,
                        quantity = 1,
                        date = date,
                        timestamp = date,
                        notes = null
                    )
                    supplementRepository.addSupplementEntry(entry)
                }
            }
        }
    }

    private fun getDemoFoods(): List<DemoFood> = listOf(
        // Breakfast
        DemoFood("Oatmeal with Berries", "Quaker", "Breakfast", 150.0, 5.0, 27.0, 3.0, 4.0, 6.0, 0.0, "1", "cup"),
        DemoFood("Greek Yogurt", "Chobani", "Breakfast", 130.0, 14.0, 11.0, 3.5, 0.0, 9.0, 65.0, "150", "g"),
        DemoFood("Scrambled Eggs", null, "Breakfast", 180.0, 12.0, 2.0, 14.0, 0.0, 1.0, 180.0, "2", "eggs"),
        DemoFood("Whole Wheat Toast", "Dave's Killer", "Breakfast", 110.0, 5.0, 22.0, 1.5, 5.0, 5.0, 170.0, "1", "slice"),
        DemoFood("Smoothie Bowl", null, "Breakfast", 280.0, 8.0, 45.0, 10.0, 6.0, 25.0, 45.0, "1", "bowl"),

        // Lunch
        DemoFood("Grilled Chicken Salad", null, "Lunch", 320.0, 35.0, 15.0, 15.0, 5.0, 8.0, 480.0, "1", "salad"),
        DemoFood("Turkey Sandwich", "Subway", "Lunch", 280.0, 18.0, 46.0, 4.5, 5.0, 7.0, 850.0, "6", "inch"),
        DemoFood("Quinoa Bowl", null, "Lunch", 380.0, 14.0, 55.0, 12.0, 8.0, 6.0, 320.0, "1", "bowl"),
        DemoFood("Salmon Wrap", null, "Lunch", 420.0, 28.0, 35.0, 18.0, 4.0, 5.0, 620.0, "1", "wrap"),
        DemoFood("Veggie Burger", "Beyond", "Lunch", 250.0, 20.0, 20.0, 14.0, 3.0, 3.0, 390.0, "1", "patty"),

        // Dinner
        DemoFood("Grilled Salmon", null, "Dinner", 367.0, 40.0, 0.0, 22.0, 0.0, 0.0, 90.0, "6", "oz"),
        DemoFood("Chicken Stir Fry", null, "Dinner", 420.0, 32.0, 38.0, 15.0, 4.0, 12.0, 890.0, "1.5", "cups"),
        DemoFood("Pasta Primavera", null, "Dinner", 380.0, 12.0, 58.0, 12.0, 6.0, 8.0, 420.0, "1.5", "cups"),
        DemoFood("Beef Tacos", null, "Dinner", 450.0, 28.0, 35.0, 22.0, 5.0, 4.0, 780.0, "3", "tacos"),
        DemoFood("Vegetable Curry", null, "Dinner", 320.0, 10.0, 48.0, 12.0, 8.0, 10.0, 580.0, "1.5", "cups"),

        // Snacks
        DemoFood("Apple", null, "Snack", 95.0, 0.5, 25.0, 0.3, 4.0, 19.0, 2.0, "1", "medium"),
        DemoFood("Protein Bar", "Quest", "Snack", 190.0, 20.0, 22.0, 8.0, 14.0, 1.0, 140.0, "1", "bar"),
        DemoFood("Mixed Nuts", "Planters", "Snack", 170.0, 6.0, 7.0, 15.0, 3.0, 1.0, 95.0, "28", "g"),
        DemoFood("Banana", null, "Snack", 105.0, 1.3, 27.0, 0.4, 3.0, 14.0, 1.0, "1", "medium"),
        DemoFood("Dark Chocolate", "Lindt", "Snack", 155.0, 1.4, 17.0, 9.0, 2.0, 14.0, 4.0, "30", "g")
    )

    private fun getDemoExercises(): List<DemoExercise> = listOf(
        DemoExercise("Running", "Cardio", 10.0, 15, 45, true),
        DemoExercise("Cycling", "Cardio", 8.0, 20, 60, true),
        DemoExercise("Swimming", "Cardio", 11.0, 20, 45, true),
        DemoExercise("Weight Training", "Strength", 6.0, 30, 60, false),
        DemoExercise("Yoga", "Flexibility", 3.0, 30, 90, false),
        DemoExercise("Walking", "Cardio", 4.0, 20, 60, true),
        DemoExercise("HIIT", "Cardio", 12.0, 15, 30, false),
        DemoExercise("Pilates", "Flexibility", 4.0, 30, 60, false),
        DemoExercise("Boxing", "Cardio", 9.0, 20, 45, false),
        DemoExercise("Dancing", "Cardio", 7.0, 30, 60, false)
    )

    private fun getDemoSupplements(): List<DemoSupplement> = listOf(
        DemoSupplement("Vitamin D3", "Nature Made", 2000.0, "IU"),
        DemoSupplement("Omega-3", "Nordic Naturals", 1000.0, "mg"),
        DemoSupplement("Multivitamin", "Centrum", 1.0, "tablet"),
        DemoSupplement("Vitamin C", "NOW Foods", 1000.0, "mg"),
        DemoSupplement("Probiotic", "Garden of Life", 50.0, "billion CFU"),
        DemoSupplement("Magnesium", "Natural Vitality", 325.0, "mg"),
        DemoSupplement("B-Complex", "Thorne", 1.0, "capsule"),
        DemoSupplement("Iron", "Solgar", 25.0, "mg"),
        DemoSupplement("Zinc", "Pure Encapsulations", 30.0, "mg"),
        DemoSupplement("Calcium", "Citracal", 630.0, "mg")
    )

    private fun Double.round(decimals: Int): Double {
        var multiplier = 1.0
        repeat(decimals) { multiplier *= 10 }
        return kotlin.math.round(this * multiplier) / multiplier
    }

    // Helper data classes
    private data class DemoFood(
        val name: String,
        val brand: String?,
        val category: String,
        val calories: Double,
        val protein: Double,
        val carbs: Double,
        val fat: Double,
        val fiber: Double?,
        val sugar: Double?,
        val sodium: Double?,
        val servingSize: String,
        val servingUnit: String
    )

    private data class DemoExercise(
        val name: String,
        val category: String,
        val caloriesPerMinute: Double,
        val minDuration: Int,
        val maxDuration: Int,
        val hasDistance: Boolean
    )

    private data class DemoSupplement(
        val name: String,
        val brand: String,
        val dosage: Double,
        val unit: String
    )
}